backend/src/services/certificate-authority/azure-ad-cs/azure-ad-cs-certificate-authority-fns.ts

```
@@ -0,0 +1,1103 @@
+};
+
+const buildSubjectDN = (commonName: string, properties?: TPkiSubscriberProperties): string => {
+  // Validate and sanitize common name - it's required and cannot be empty
+  if (!commonName || !commonName.trim()) {
+    throw new BadRequestError({ message: "Common Name is required and cannot be empty" });
+  }
+
+  const sanitizedCN = commonName.trim().replace(new RE2("[,=+<>#;\\\\\\]]", "g"), "");
+  if (!sanitizedCN) {
+    throw new BadRequestError({ message: "Common Name contains only invalid characters" });
+  }
+
+  let subject = `CN=${sanitizedCN}`;
+
+  // Helper function to validate and sanitize DN component values
+  const sanitizeComponent = (value: string | undefined): string | null => {
+    if (!value || typeof value !== "string") return null;
+    const trimmed = value.trim();
+    if (!trimmed) return null;
+
+    // Remove or escape problematic characters for DN components
+    // Be more aggressive in removing characters that can cause OID issues
+    const sanitized = trimmed.replace(new RE2('[,=+<>#;\\\\"\\/\\r\\n\\t]', "g"), "").trim();
+
+    // Additional validation to prevent empty components that cause OID errors
+    if (sanitized.length === 0) return null;
+
+    // Ensure the component doesn't start or end with spaces or problematic chars
+    const finalSanitized = sanitized.replace(new RE2("^[\\\\s\\\\-_.]+|[\\\\s\\\\-_.]+$", "g"), "");
+
+    return finalSanitized.length > 0 ? finalSanitized : null;
+  };
+
+  // Build DN components in proper order for ADCS compatibility
+  // Order matters for Microsoft ADCS - follow X.500 standard ordering
+
+  const emailAddress = sanitizeComponent(properties?.emailAddress);
+  if (emailAddress) {
+    // Enhanced email validation for DN usage
+    const emailRegex = new RE2(/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/);
+    if (emailRegex.test(emailAddress) && emailAddress.length > 5 && emailAddress.length < 64) {
+      subject += `,E=${emailAddress}`;
+    }
```

should we throw an error if it doesn't match rather than silently ignoring?

---

I think ideally it should be properly propagated to the user what the issues are with their inputs. Maybe we throw this upon subscriber create/update?